RFCs/symbol_alias.md

```diff
@@ -0,0 +1,171 @@
+# Symbol aliases
+
+## Abstract
+
+This RFC proposes to add more aliasing capabilties to Nim. Every symbol (template, module, const, iterator etc: anything in `TSymKind`) becomes a 1st class entity, which can be passed to routines (procs, templates, iterators, etc: anything in `routineKinds`), or used as generic paramter in types. This works even if the symbol is overloaded, or if it's a template/macro (even if all parameters are optional).
+
+In particular, this allows defining 0-cost lambdas and aliases.
+
+See https://github.com/nim-lang/Nim/pull/11992 for the corresponding PR.
+
+## Motivation 1: symbol aliasing
+* nim doesn't have a way to alias symbols, and the workaround using `template foo2(args: varargs[untyped]): untyped = foo(args)` falls short in many cases; eg
+  - if all parameters are optional, this won't work, see `case1`
+  - this requires different syntax for different symbols and doesn't work for some symbols, see `case2`
+  - it only provides an alternative way to call a routine but doesn't alias the symbol, so for example using introspection (eg to get parameter names) would give different results, see `case3`
+
+## Motivation 2: passing symbols to routines
+
+* nim doesn't allow passing templates or uninstantiated generics (or macros, iterators etc) to routines, so it forces you to either turn the callee into a template/macro, leading to much more complex, hard to debug code code (eg: see zero-functional https://github.com/zero-functional/zero-functional/blob/master/zero_functional.nim, with macros calling macros calling macros...).
+
+## Motivation 3: speed benefit
+Passing functors via closures is not free, because closure prevents inlining. Instead this RFC allows passing templates directly which is a zero-cost abstraction. eg: see `case4`.
+
+## Motivation 3: lambdas
+* this allows defining lambdas, which are more flexible than `sugar.=>`, and don't incur overhead, see `tests/magics/tlambda.nim`; eg: `case5`
+
+## Motivation 4: composable iterators
+composable iterators, which allows define lazy functional programming primitives (eager `toSeq` and lazy `map/filter/join` etc; which can have speed benefit compared to `sequtils`); the resulting code is quite nice and compact, see `tlambda_iterators.nim` which uses library `lambdaIter` that builds on top of `lambda`; it allows defining primitives that work regardless we're passing a value (`@[1,2,3]`) or a lazy iterate (eg `iota(3)`)
+
+## Motivation 5: this fixes or closes a lot of issues
+see a sample here: https://github.com/nim-lang/Nim/pull/11992
```

Comment:
It's logically impossible that a feature addition fixes existing bugs. What do you really mean? That the patch also happens to fix existing compiler bugs? That the new feature allows for new workarounds?

Reply:
the distinction between bugs and feature requests is by definition a gray area, and some github issues are reducible to feature requests so there's no contradiction. 
Anything I listed in https://github.com/nim-lang/Nim/pull/11992 as closable should be closable with that PR (because it'd make no sense keeping the issue opened past that); but of course we can debate that list 1 by 1.

* in some cases because this PR provides a good enough solution (I wouldn't call it workaroud unless there's some kind of drawback involved)
* in other cases it's a proper fix, eg:
  - https://github.com/nim-lang/Nim/issues/4516 "it should be possible to pass an inline iterator to another inline iterator." (OP doesn't require a specific syntax)
  - https://github.com/nim-lang/Nim/issues/11870 "symbols of no argument templates can't be passed to macros" => OP doesn't impose a specific syntax; this PR proposes a general fix for that so it fixes this issue as well
  - https://github.com/nim-lang/Nim/issues/7435 "type inference with lambdas doesn't work (requires explicit types)"
likewise, that issue doesn't care whether it's `sugar.=>` or something else, and this PR solves this via `~>` (with plenty of tests)
  - https://github.com/nim-lang/Nim/issues/7090 "alias declarations" (was marked as closed but it should not have been closed for reasons explained in this PR+RFC; this PR properly fixes it)

and IIRC the PR has tests for all issues I intend to close with this PR.